examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,267 @@
+/* This test based on the following pseudo code and tests correctness of our
+ * new allocator snmalloc
+
+   for 0..num_threads {
+    loop {
+        let mut regions: Vec<(Box<[u8], u8>)>;
+        let mut mem_used = 0u64;
+        match rnd % 4 {
+           0 => // Check area
+           1..2 => // Alloc random area if less than x% of the heap is used (not 100% to
+                   // account for fragmentation, ...) and check area.
+           3 => // Free random area
+        }
+    }
+}
+ * So this basically runs for a long time and should never crash
+ */
+
+
+use core::arch::asm;
+use num_cpus;
+use rand::Rng;
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread;
+use std::time::Instant;
+
+extern "C" {
+    static HEAP_BASE: u64;
+    static HEAP_SIZE: usize;
+}
+
+const PAGE_SIZE: usize = 4096;
+const TO_KB: usize = 1024;
+const TO_MB: usize = TO_KB * 1024;
+const TO_GB: usize = TO_MB * 1024;
+
+const NUM_OPERATION_CHOICES: usize = 4;
+const MAGIC_NUM_TO_CHECK: u32 = 0xdeadbeef;
```

Comment:
I have a slight preference for not using `0xdeadbeef` as a magic number. This is a popular choice for magic number, so if we're using 3rd party libraries, there's a slightly higher chance of this particular magic number being used by something else, compared to picking some different magic number. Also, 32 bits is pretty small if it's important that we don't encounter the magic number on accident. 64 bits would be better, if that's possible.

Both of these comments are very minor.